File: GFG/Sorting/countinversions.py
def merge(arr,l,m,r):
    n1=m-l+1
    n2=r-m
    left=[]
    right=[]
    for i in range(n1):
        left.append(arr[l+i])
    for i in range(n2):
        right.append(arr[m+1+i])
    i=0
    j=0
    k=l
    while(i<n1 and j<n2):
        if(left[i]<right[j]):
            arr[k]=left[i]
            i+=1
            k+=1
        else:
            arr[k]=right[j]
            j+=1
            k+=1
            global count
            count+=(n1-i)
    while(i<n1):
        arr[k]=left[i]
        k+=1
        i+=1
    while (j<n2):
        arr[k]=right[j]
        j+=1
        k+=1

def mergesort(arr,l,r):
    if(r>l):
        m=l+(r-l)//2
        mergesort(arr,l,m)
        mergesort(arr,m+1,r)
        merge(arr,l,m,r)
        
arr=[5,10,10,13,8,9,14,34,7,7,19,99]
count=0
mergesort(arr,0,len(arr)-1)
print(arr)
print(count)

File: GFG/Sorting/countinversions.cpp
#include<iostream>
#include<vector>
using namespace std;

int cntnmerge(vector<int> &arr,int l, int m, int r){
    int n1=m-l+1;
    int n2=r-m;
    vector<int> left(n1),right(n2);
    for(int i=0;i<n1;i++){
        left[i]=arr[l+i];
    }
    for(int i=0;i<n2;i++){
        right[i]=arr[m+i+1];
    }
    int i=0;
    int j=0;
    int k=l;
    int res=0;
    while(i<n1 && j<n2){
        if(left[i]<right[j]){
            arr[k]=left[i];
            i++;
            k++;
        }
        else{
            arr[k]=right[j];
            k++;
            j++;
            res=res+n1-i;
        }
    }
    while(i<n1){
        arr[k]=left[i];
        i++;
        k++;
    }
    while(j<n2){
        arr[k]=right[j];
        k++;
        j++;
    }
    return res;
}

int cntinv(vector<int> &arr,int l,int r){
    int res=0;
    if(r>l){
        
        int m=l+(r-l)/2;
        res+=cntinv(arr,l,m);
        res+=cntinv(arr,m+1,r);
        res+=cntnmerge(arr,l,m,r);
    }
    return res;
}
int main(){
    vector<int> arr = {5,10,10,13,8,9,14,34,7,7,19,99};

    cout << "Original array: ";
    for (int num : arr)
        cout << num << " ";
    cout << endl;

    int result=cntinv(arr, 0, arr.size() - 1);

    cout << "Sorted array: ";
    for (int num : arr)
        cout << num << " ";
    cout << endl;
    cout<<result;

}